### DISCUSSION TITLE
How can we provide content for specific links

### DISCUSSION THREAD

**Discussion by Mister-Hope** [vuepress/core#145] [Q&A] [answered]
In V1, we can use `beforeDevServer` to provide content for specific  links, e.g: 

Providing a manifest file:

```ts
{
    beforeDevServer(app) {
      app.get(`${base || "/"}manifest.webmanifest`, (_req, res) => {
        getManifest(pwaOption, app)
          .then((manifest) => {
            res.send(manifest);
          })
          .catch(() =>
            res.status(500).send("Unexpected manifest generate error")
          );
      });
    },
}
```

How can we do that now after beforeDevServer is removed?

---

The scene describe above is quite common for plugins which generate files during the build process. These plugins should provide content while the devServer is running for the best experience.

**Comment by meteorlxy** [comment, marked as answer]
If you want it anyway:

```js
  // for vite
  if (app.env.isDev && app.options.bundler.endsWith('vite')) {
    app.options.bundlerConfig.viteOptions = require('vite').mergeConfig(
      app.options.bundlerConfig.viteOptions,
      {
        // ...
      }
    )
  }

  // for webpack
  if (app.env.isDev && app.options.bundler.endsWith('webpack')) {
    const currentBeforeDevServer = app.options.bundlerConfig.beforeDevServer

	app.options.bundlerConfig.beforeDevServer = (...args) => {
		// ...
		currentBeforeDevServer(...args)
	}
  }
```

**Reply by Mister-Hope** [reply]
Got it, thx, while I still think adding some support in the apis would be better. I think it's possible to build a wrapper to fit the existing two bundler, and make them both work with this wrapped function helper as a new API.😃

I have to check vite docs if I want this feature now because I am not familiar with vite, and it's a bit hard. I have 3 plugins depending on this feature and I do think it's common for the case that a plugin providing content in devServer which will be added after buildprocess (like sitemap, manifest, rss, feed and so on). It will improve users experience if they can see the reactive results in devServer
